feat: add keyboard suport for the canvas view

rthor opened the pull request:
### Why?

I can't use a mouse/trackpad with scroll functionality and therefore need keyboard support to be able to view and navigate large statecharts.

### What this PR does:
Adds keyboard support for:
- Panning
- Zooming
- Fit to view
- Reset view

![Screen Shot 2021-08-31 at 12 06 20](https://user-images.githubusercontent.com/1608329/131499593-fb55499c-89fc-4f53-b5e3-a6a60e97d733.png)


### Note to maintainers:

I already talked with @davidkpiano about needing this functionality, and I hope it's okay that I just went ahead and added it. Wanted to use the new visualizer ASAP :) it's such a great workflow improvement for dealing with xstate! If you want this done differently, I am happy to help, or otherwise close this PR.

And congratz on the release 🎉

mattpocock commented:
@rthor Very nice! Spotted some odd behaviour in the top-right when you're on the tabs:

https://www.loom.com/share/25d827205468451b9fb84ec5636a418f

mattpocock commented on src/CanvasContainer.tsx:
```diff
@@ -97,6 +97,34 @@ const dragMachine = dragModel.createMachine({
+  if (positive.includes(key)) {
+    return delta;
+  } else if (negative.includes(key)) {
+    return -delta;
```
Just a nit, feel free to ignore

mattpocock commented:
An idea for a fix for the above would be not allowing WASD navigation if the element you're focusing has a role of 'tab'. Kinda weird, maybe there's a better way of doing this @Andarist?

```ts
export const isTabLikeElement = (el: HTMLElement) => {
  return el.getAttribute('role') === 'tab';
};
```

Andarist reviewed (changes requested):
Thank you for an excellent PR! I have left some nits and requested minor changes. I've tested this though and the core of the functionality looks great. It would be nice to get some tests for this - but it's definitely more important for this to be merged in sooner than later so if you don't have time to figure out how to write Cypress tests for this - it's OK. We should add them eventually :p 

BTW. I'm also working today on the Hand icon so you could switch to the Pan mode when that gets pressed. I hope this will also be a nice improvement in this area.

Andarist commented:
> An idea for a fix for the above would be not allowing WASD navigation if the element you're focusing has a role of 'tab'. Kinda weird, maybe there's a better way of doing this @Andarist?

I don't really have a better idea than this - it's just a bit of a problem because there are a lot of default interactions that we could forget about and logic like this should, ideally, not conflict with those. The only two approaches I can think of for handling this is to:
- try to prevent activating the logic based on conditions like this one
- try to stop propagation everywhere

The problem with the second approach is that it's hard to apply it globally and it requires adding this almost everywhere - while also requiring listeners for native button behavior ("clicking" using Enter/Space), etc. So it, IMHO, doesn't scale. This makes me stick with the first approach - although it has its shortcomings as well.

#usetheplatform

rthor commented on src/CanvasContainer.tsx:
```diff
@@ -215,6 +243,43 @@ export const CanvasContainer: React.FC = ({ children }) => {
+      const dx = getDeltaX(e.key, e.shiftKey);
+      const dy = getDeltaY(e.key, e.shiftKey);
+
+      if (e.key === '+') {
```
In my experience `+` seems to be more common, but I have no preference, and am happy to do whichever you prefer.

Andarist replied:
Let's stick to `+` then 👍  I definitely have no preference of my own here

rthor commented on src/CanvasContainer.tsx:
```diff
@@ -215,6 +243,43 @@ export const CanvasContainer: React.FC = ({ children }) => {
+  useEffect(() => {
+    function keydownListener(e: KeyboardEvent) {
+      const target = e.target as HTMLElement;
+      if (isTextInputLikeElement(target)) {
```
So attaching the listeners directly to the canvas container? I was going to do this initially, only opted for global since that is being done for the spacebar lock logic.

Andarist replied:
It's often easy to focus `body` or some other element - so hooking into a global thing like `window` seems to be preferable because we won't miss keystrokes. 

Sorry if my question was confusing - the question was about if conceptually those kinds of things are truly global, which includes those being triggered also on activatable elements. It was sort of an introduction to the comment about the range input because I've wanted to establish what kind of filtering logic we need to support.

rthor commented:
Thanks for the quick reviews! I'll go through those and make appropriate fixes. 

Regarding the control logic for panning, would these types of events make sense?

```tsx
{
  on: {
    'PAN.LEFT': {
      actions: canvasModel.assign({
        pan: (ctx, e) => ({
          dx: ctx.pan.dx - (e.isLongPan ? LONG_PAN : SHORT_PAN),
          dy: ctx.pan.dy,
        }),
      }),
      target: '.throttling',
      internal: false,
    },
    'PAN.RIGHT': {
      // ...
    },
    'PAN.UP': {
      // ...
    },
    'PAN.DOWN': {
      // ...
    },
}
```

Then for choosing the correct event per key, would we just like a switch case:

```tsx
switch (e.key) {
    case 'a':
    case 'ArrowLeft':
      return canvasService.send('PAN.LEFT')
    case 'd':
    // ...
}
```

Or is there a better way to do this?

Andarist commented:
If we want to move more of this logic to the machine itself then the proposed changes sound OK to me.

rthor commented:
Moved the logic to canvasMachine, increased the small pan delta, and fixed the tab behavior with your proposed solution @mattpocock. Testing this should be quite straightforward. Can either do it in this PR or follow up. Just don't have time right now. 

> BTW. I'm also working today on the Hand icon so you could switch to the Pan mode when that gets pressed. I hope this will also be a nice improvement in this area.

Looking forward to this!!

Andarist commented on src/CanvasContainer.tsx:
```diff
@@ -215,6 +219,64 @@ export const CanvasContainer: React.FC = ({ children }) => {
+    function keydownListener(e: KeyboardEvent) {
+      const target = e.target as HTMLElement;
+
+      if (isTextInputLikeElement(target) || isTabLikeElement(target)) {
```
This is, of course, nit-picking (to some extent) but it seems to me that if we want to have a bullet-proof logic that works in the very same way for different interactions then we should **selectively** apply those filters based on the pressed key.

Why this is IMHO important? If we allow, let's say, WSAD keys on buttons then why we don't allow them on tab elements? Tab elements don't have WSAD-based interactions of their own, after all. 

In addition to that - we should not handle arrow keys here when the target is an input of type range.

Andarist commented on src/utils.ts:
```diff
@@ -210,3 +210,6 @@ export const isTextInputLikeElement = (el: HTMLElement) => {
 };
+
+export const isTabLikeElement = (el: HTMLElement) =>
+  el.getAttribute('role') === 'tab';
```
An interesting fact is that `@testing-library/dom` cheks all space-delimited roles:
https://github.com/testing-library/dom-testing-library/blob/fbbb29a6d9655d41bc8f91d49dc64326f588c0d6/src/queries/role.js#L107-L112

Should we use the same logic for that? 🤔 

nit: this probably should be called `isTabElement` or `hasTabRole`

Andarist replied:
We'll probably have the very same issue with arrows triggering canvas interactions when navigating through menus. Damn, handling all the cases here is hard! 😬

farskid commented:
@rthor Testing this now and I realized all the keybindings are working except for zooming.
Pressing `-` zooms out but zooming in happens only via `Shift` and `+`.

farskid reviewed (changes requested):
@rthor Testing this now and I realized all the keybindings are working except for zooming.
Pressing `-` zooms out but zooming in happens only via `Shift` and `+`.

I'm on the latest Chrome.

Andarist commented:
@farskid see the comment here: https://github.com/statelyai/xstate-viz/pull/237#discussion_r699341260 . I'm OK with adjusting this however we want, I just have not researched this enough. For instance, Excalidraw requires Shift for *both* zooming out and zooming in

farskid commented:
> @farskid see the comment here: https://github.com/statelyai/xstate-viz/pull/237#discussion_r699341260 . I'm OK with adjusting this however we want, I just have not researched this enough. For instance, Excalidraw requires Shift for *both* zooming out and zooming in

My issue with it is that why we need shift for one when the other one doesn't

mattpocock reviewed (approved):
Looks great, fit-to-view is working. The +, - thing doesn't bother me particularly, happy to approve without fiddling with that.

Andarist commented:
@farskid I've changed this to not require Shift for both zoom in/out. I've also added numpad handling 😱 since it has -/+ too
